notification/notification.go

```diff
@@ -0,0 +1,135 @@
+package notification
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+
+	"fmt"
+
+	"github.com/fabric8-services/fabric8-auth/goasupport"
+	"github.com/fabric8-services/fabric8-auth/log"
+	"github.com/fabric8-services/fabric8-auth/notification/client"
+	"github.com/fabric8-services/fabric8-auth/rest"
+	goaclient "github.com/goadesign/goa/client"
+	goauuid "github.com/goadesign/goa/uuid"
+	uuid "github.com/satori/go.uuid"
+)
+
+// Channel is a simple interface between the notifying component and the notificaiton impl
+type Channel interface {
+	Send(context.Context, Message)
+}
+
+// Message represents a new event of a Type for a Target performed by a User
+// See helper constructors like NewUserEmailUpdated
+type Message struct {
+	MessageID   uuid.UUID // unique ID per event
+	UserID      *string
+	Custom      map[string]interface{}
+	TargetID    string
+	MessageType string
+}
+
+func (m Message) String() string {
+	return fmt.Sprintf("id:%v type:%v by:%v for:%v", m.MessageID, m.MessageType, m.UserID, m.TargetID)
+}
+
+func NewUserEmailUpdated(userID string, custom map[string]interface{}) Message {
+	return Message{
+		MessageID:   uuid.NewV4(),
+		MessageType: "user.email.update",
+		TargetID:    userID,
+		UserID:      &userID, // in future if service accounts are allowed to update, this will be handy.
+		Custom:      custom,
+	}
+}
+
+// DevNullChannel is the default configured channel. It does nothing.
+type DevNullChannel struct{}
+
+// Send NO-OP
+func (d *DevNullChannel) Send(context.Context, Message) {}
+
+// ServiceConfiguration holds configuration options required to interact with the fabric8-notification API
+type ServiceConfiguration interface {
+	GetNotificationServiceURL() string
+}
+
+// Service is a simple client Channel to the fabric8-notification service
+type Service struct {
+	config ServiceConfiguration
+}
+
+func validateConfig(config ServiceConfiguration) error {
+	_, err := url.Parse(config.GetNotificationServiceURL())
+	if err != nil {
+		return fmt.Errorf("Invalid NotificationServiceURL %v cause %v", config.GetNotificationServiceURL(), err.Error())
+	}
+	return nil
+}
+
+// NewServiceChannel sends notification messages to the fabric8-notification service
+func NewServiceChannel(config ServiceConfiguration) (Channel, error) {
+	err := validateConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return &Service{config: config}, nil
+}
+
+// Send invokes the fabric8-notification API
+func (s *Service) Send(ctx context.Context, msg Message) {
+	go func(ctx context.Context, msg Message) {
+
+		u, err := url.Parse(s.config.GetNotificationServiceURL())
```

Comment:
You already parsed it once in NewServiceChannel (line 65). Why don't save it as an URL in the Service struct (instead of ServiceConfiguration) instead of re-parsing it every time we are sending a message?

Reply:
We could do that for sure - but it (kind of) deviates from how we use configuration information.
In all places, 
1. We initialize the config object in the constructor helper ( `NewService..` ).
2. We consume the content the config object in the service/struct methods.

but I'm trying to avoid deviating into initializing the config value directly inside the constructor. 
WDYT ?
Dont have too strong feelings about this :)

Reply:
No, no. You still pass the configuration to the constructor. But then get the URL from it and parse it (as you already do) then save it instead of throwing it away. Then just reuse it every time you need it instead of re-parsing over and over again every time.

Reply:
Sure!